### DISCUSSION TITLE
Non-linear Kernel Integration

### DISCUSSION THREAD

**Discussion by dmadisetti** [quarto-dev/quarto-cli#9362] [Q&A]
### Description

I've started replacing my jupyter notebooks with [marimo](https://github.com/marimo-team/marimo) which honestly feels so much better.
The big selling point / difference of marimo is its reactive nature (think observable but for python).
As far as I understand, this makes a kernel a little difficult with quarto- since marimo needs to ingest the whole notebook before running and the cells are run as a DAG.

I got this working with 2 filter passes, with a background server process, but the server seems like a bit of a hack: https://github.com/dmadisetti/quarto-marimo

```mermaid
graph TD
    A[run.sh] --Spawns (server.py)--> B[Flask server]
    A --> C[Runs Quarto]
    C --> F1
    subgraph F1[Filter 1: marimo-register]
        D[`marimo-register:CodeBlock`]
        E[`marimo-register:Pandoc`]
    end
    D --sends cells to build--> B[Builds all the cells]
    D --replaces: `{marimo}` code blocks with {marimo-id} stubs--> E
    E --Signals server to run on cell objects--> B
    F1 --> F2
    subgraph F2[Filter 2: marimo-execute]
        G[`marimo-execute:CodeBlock`]
        H[`marimo-execute:Pandoc`]
    end
    B --Sends code outputs--> G
    G --Replaces stubs with outputs--> H
    H --Provides relevant headers to make things work--> Out
```

Wondering if anyone had insight on how to have proper integration

**Comment by mcanouil** [comment]
Does the "interactive" part requires to have a running Python session?

**Reply by dmadisetti** [reply]
It can, but marimo also works with WASM- which would be my preferred use case here; just using Quarto for the preprocess
